database/repository/candle/candle.go

```diff
@@ -0,0 +1,273 @@
+		qm.Where("timestamp between ? and ?", start.UTC(), end.UTC()),
+	}
+
+	exchangeUUID, errS := exchange.UUIDByName(exchangeName)
+	if errS != nil {
+		return out, errS
+	}
+	queries = append(queries, qm.Where("exchange_name_id = ?", exchangeUUID.String()))
+
+	if repository.GetSQLDialect() == database.DBSQLite3 {
+		retCandle, errC := modelSQLite.Candles(queries...).All(context.Background(), database.DB.SQL)
+		if errC != nil {
+			return out, errC
+		}
+		for x := range retCandle {
+			t, errT := time.Parse(time.RFC3339, retCandle[x].Timestamp)
+			if errT != nil {
+				return out, errT
+			}
+			out.Candles = append(out.Candles, Candle{
+				Timestamp: t,
+				Open:      retCandle[x].Open,
+				High:      retCandle[x].High,
+				Low:       retCandle[x].Low,
+				Close:     retCandle[x].Close,
+				Volume:    retCandle[x].Volume,
+			})
+		}
+	} else {
+		retCandle, errC := modelPSQL.Candles(queries...).All(context.Background(), database.DB.SQL)
+		if errC != nil {
+			return out, errC
+		}
+
+		for x := range retCandle {
+			out.Candles = append(out.Candles, Candle{
+				Timestamp: retCandle[x].Timestamp,
+				Open:      retCandle[x].Open,
+				High:      retCandle[x].High,
+				Low:       retCandle[x].Low,
+				Close:     retCandle[x].Close,
+				Volume:    retCandle[x].Volume,
+			})
+		}
+	}
+	if len(out.Candles) < 1 {
+		return out, fmt.Errorf(errNoCandleDataFound, exchangeName, base, quote, interval, asset)
+	}
+
+	out.ExchangeID = exchangeName
+	out.Interval = interval
+	out.Base = base
+	out.Quote = quote
+	out.Asset = asset
+	return out, err
+}
+
+// Insert series of candles
+func Insert(in *Item) (uint64, error) {
+	if database.DB.SQL == nil {
+		return 0, database.ErrDatabaseSupportDisabled
+	}
+
+	ctx := context.Background()
+	tx, err := database.DB.SQL.BeginTx(ctx, nil)
+	if err != nil {
+		return 0, err
+	}
+
+	var totalInserted uint64
+	if repository.GetSQLDialect() == database.DBSQLite3 {
+		totalInserted, err = insertSQLite(ctx, tx, in)
+	} else {
+		totalInserted, err = insertPostgresSQL(ctx, tx, in)
+	}
+	if err != nil {
+		return 0, err
+	}
+
+	err = tx.Commit()
+	if err != nil {
+		errRB := tx.Rollback()
+		if errRB != nil {
+			log.Errorln(log.DatabaseMgr, errRB)
+		}
+		return 0, err
+	}
+	return totalInserted, nil
+}
+
+func insertSQLite(ctx context.Context, tx *sql.Tx, in *Item) (uint64, error) {
+	var totalInserted uint64
+	for x := range in.Candles {
+		var tempCandle = modelSQLite.Candle{
+			ExchangeNameID: in.ExchangeID,
+			Base:           in.Base,
+			Quote:          in.Quote,
+			Interval:       strconv.FormatInt(in.Interval, 10),
+			Asset:          in.Asset,
+			Timestamp:      in.Candles[x].Timestamp.UTC().Format(time.RFC3339),
+			Open:           in.Candles[x].Open,
+			High:           in.Candles[x].High,
+			Low:            in.Candles[x].Low,
+			Close:          in.Candles[x].Close,
+			Volume:         in.Candles[x].Volume,
+		}
+		tempUUID, err := uuid.NewV4()
+		if err != nil {
+			return 0, err
```

Comment:
With this here, it means you can still encounter the transaction being stuck issue. I don't expect `newv4` to fail, but that potential is there

Reply:
Yeah I moved the Rollback() call back out of the insertSQLite() & insertPostgresSQL() calls had this earlier but apparently i reverted it again 🤷 

This will handle a Rollback on any insert errors